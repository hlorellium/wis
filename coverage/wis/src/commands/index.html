
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for wis/src/commands</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> wis/src/commands</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">85.64% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>340/397</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">94.73% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>72/76</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">79.62% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>43/54</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">85.64% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>340/397</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="deleteShapeCommand.ts"><a href="deleteShapeCommand.ts.html">deleteShapeCommand.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="42" class="abs high">42/42</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="42" class="abs high">42/42</td>
	</tr>

<tr>
	<td class="file medium" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="75" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 75%"></div><div class="cover-empty" style="width: 25%"></div></div>
	</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="164" class="abs medium">123/164</td>
	<td data-value="92.85" class="pct high">92.85%</td>
	<td data-value="28" class="abs high">26/28</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="33" class="abs medium">22/33</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="164" class="abs medium">123/164</td>
	</tr>

<tr>
	<td class="file high" data-value="moveShapesCommand.ts"><a href="moveShapesCommand.ts.html">moveShapesCommand.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="85" class="abs high">85/85</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="21" class="abs high">21/21</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="85" class="abs high">85/85</td>
	</tr>

<tr>
	<td class="file high" data-value="moveVertexCommand.ts"><a href="moveVertexCommand.ts.html">moveVertexCommand.ts</a></td>
	<td data-value="84.9" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 84%"></div><div class="cover-empty" style="width: 16%"></div></div>
	</td>
	<td data-value="84.9" class="pct high">84.9%</td>
	<td data-value="106" class="abs high">90/106</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="16" class="abs high">14/16</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="84.9" class="pct high">84.9%</td>
	<td data-value="106" class="abs high">90/106</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-06-29T17:08:27.657Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    