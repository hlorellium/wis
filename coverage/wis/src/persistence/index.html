
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for wis/src/persistence</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> wis/src/persistence</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">68.37% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>253/370</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">67.08% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>53/79</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">82.85% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>29/35</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">68.37% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>253/370</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="indexedDbStore.ts"><a href="indexedDbStore.ts.html">indexedDbStore.ts</a></td>
	<td data-value="65.78" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 65%"></div><div class="cover-empty" style="width: 35%"></div></div>
	</td>
	<td data-value="65.78" class="pct medium">65.78%</td>
	<td data-value="152" class="abs medium">100/152</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="30" class="abs medium">18/30</td>
	<td data-value="73.68" class="pct medium">73.68%</td>
	<td data-value="19" class="abs medium">14/19</td>
	<td data-value="65.78" class="pct medium">65.78%</td>
	<td data-value="152" class="abs medium">100/152</td>
	</tr>

<tr>
	<td class="file medium" data-value="persistenceManager.ts"><a href="persistenceManager.ts.html">persistenceManager.ts</a></td>
	<td data-value="70.18" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 70%"></div><div class="cover-empty" style="width: 30%"></div></div>
	</td>
	<td data-value="70.18" class="pct medium">70.18%</td>
	<td data-value="218" class="abs medium">153/218</td>
	<td data-value="71.42" class="pct medium">71.42%</td>
	<td data-value="49" class="abs medium">35/49</td>
	<td data-value="93.75" class="pct high">93.75%</td>
	<td data-value="16" class="abs high">15/16</td>
	<td data-value="70.18" class="pct medium">70.18%</td>
	<td data-value="218" class="abs medium">153/218</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-06-29T16:57:19.548Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    