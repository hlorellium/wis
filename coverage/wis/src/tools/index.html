
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for wis/src/tools</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> wis/src/tools</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">67.3% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>599/890</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">80.62% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>154/191</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">84.9% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>45/53</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">67.3% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>599/890</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="drawingTools.ts"><a href="drawingTools.ts.html">drawingTools.ts</a></td>
	<td data-value="25.98" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 25%"></div><div class="cover-empty" style="width: 75%"></div></div>
	</td>
	<td data-value="25.98" class="pct low">25.98%</td>
	<td data-value="177" class="abs low">46/177</td>
	<td data-value="53.33" class="pct medium">53.33%</td>
	<td data-value="15" class="abs medium">8/15</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="8" class="abs medium">6/8</td>
	<td data-value="25.98" class="pct low">25.98%</td>
	<td data-value="177" class="abs low">46/177</td>
	</tr>

<tr>
	<td class="file high" data-value="editTool.ts"><a href="editTool.ts.html">editTool.ts</a></td>
	<td data-value="80.27" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 80%"></div><div class="cover-empty" style="width: 20%"></div></div>
	</td>
	<td data-value="80.27" class="pct high">80.27%</td>
	<td data-value="370" class="abs high">297/370</td>
	<td data-value="82.41" class="pct high">82.41%</td>
	<td data-value="91" class="abs high">75/91</td>
	<td data-value="86.66" class="pct high">86.66%</td>
	<td data-value="15" class="abs high">13/15</td>
	<td data-value="80.27" class="pct high">80.27%</td>
	<td data-value="370" class="abs high">297/370</td>
	</tr>

<tr>
	<td class="file medium" data-value="panTool.ts"><a href="panTool.ts.html">panTool.ts</a></td>
	<td data-value="54.54" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 54%"></div><div class="cover-empty" style="width: 46%"></div></div>
	</td>
	<td data-value="54.54" class="pct medium">54.54%</td>
	<td data-value="55" class="abs medium">30/55</td>
	<td data-value="62.5" class="pct medium">62.5%</td>
	<td data-value="8" class="abs medium">5/8</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="6" class="abs high">5/6</td>
	<td data-value="54.54" class="pct medium">54.54%</td>
	<td data-value="55" class="abs medium">30/55</td>
	</tr>

<tr>
	<td class="file medium" data-value="selectTool.ts"><a href="selectTool.ts.html">selectTool.ts</a></td>
	<td data-value="70.1" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 70%"></div><div class="cover-empty" style="width: 30%"></div></div>
	</td>
	<td data-value="70.1" class="pct medium">70.1%</td>
	<td data-value="184" class="abs medium">129/184</td>
	<td data-value="82.97" class="pct high">82.97%</td>
	<td data-value="47" class="abs high">39/47</td>
	<td data-value="86.66" class="pct high">86.66%</td>
	<td data-value="15" class="abs high">13/15</td>
	<td data-value="70.1" class="pct medium">70.1%</td>
	<td data-value="184" class="abs medium">129/184</td>
	</tr>

<tr>
	<td class="file high" data-value="toolManager.ts"><a href="toolManager.ts.html">toolManager.ts</a></td>
	<td data-value="93.26" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 93%"></div><div class="cover-empty" style="width: 7%"></div></div>
	</td>
	<td data-value="93.26" class="pct high">93.26%</td>
	<td data-value="104" class="abs high">97/104</td>
	<td data-value="90" class="pct high">90%</td>
	<td data-value="30" class="abs high">27/30</td>
	<td data-value="88.88" class="pct high">88.88%</td>
	<td data-value="9" class="abs high">8/9</td>
	<td data-value="93.26" class="pct high">93.26%</td>
	<td data-value="104" class="abs high">97/104</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-06-29T16:50:12.950Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    