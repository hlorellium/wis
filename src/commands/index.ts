import type { State, Shape } from '../state';

export interface Command {
    apply(state: State): void;   // execute (redo)
    invert(state: State): void;  // undo
    merge?(other: Command): Command | null; // optional command merging
}

export class AddShapeCommand implements Command {
    private readonly shape: Shape;

    constructor(shape: Shape) {
        // Clone the shape to remove any proxy wrappers that can't be serialized
        this.shape = JSON.parse(JSON.stringify(shape));
    }

    apply(state: State): void {
        state.scene.shapes.push(this.shape);
    }

    invert(state: State): void {
        state.scene.shapes = state.scene.shapes.filter(sh => sh.id !== this.shape.id);
    }

    serialize(): { shape: Shape } {
        return { shape: this.shape };
    }
}

export class RemoveShapeCommand implements Command {
    private readonly shape: Shape;

    constructor(shape: Shape) {
        // Clone the shape to remove any proxy wrappers that can't be serialized
        this.shape = JSON.parse(JSON.stringify(shape));
    }

    apply(state: State): void {
        state.scene.shapes = state.scene.shapes.filter(sh => sh.id !== this.shape.id);
    }

    invert(state: State): void {
        state.scene.shapes.push(this.shape);
    }

    serialize(): { shape: Shape } {
        return { shape: this.shape };
    }
}

export class PanCommand implements Command {
    private readonly dx: number;
    private readonly dy: number;

    constructor(dx: number, dy: number) {
        this.dx = dx;
        this.dy = dy;
    }

    apply(state: State): void {
        state.view.panX += this.dx;
        state.view.panY += this.dy;
    }

    invert(state: State): void {
        state.view.panX -= this.dx;
        state.view.panY -= this.dy;
    }

    merge(other: Command): Command | null {
        if (other instanceof PanCommand) {
            return new PanCommand(this.dx + other.dx, this.dy + other.dy);
        }
        return null;
    }

    serialize(): { dx: number; dy: number } {
        return { dx: this.dx, dy: this.dy };
    }
}
