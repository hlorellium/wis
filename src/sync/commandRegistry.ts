import type { Command } from '../commands';
import { AddShapeCommand, RemoveShapeCommand, PanCommand, UndoCommand, RedoCommand, DeleteShapeCommand } from '../commands';
import type { Shape } from '../state';

export interface SerializableCommand {
    serialize(): any;
}

export interface CommandFactory {
    deserialize(data: any): Command;
}

/**
 * Registry for command serialization and deserialization
 */
export class CommandRegistry {
    private static factories = new Map<string, CommandFactory>();

    static register(commandName: string, factory: CommandFactory): void {
        this.factories.set(commandName, factory);
    }

    static deserialize(type: string, data: any): Command {
        const factory = this.factories.get(type);
        if (!factory) {
            throw new Error(`No factory registered for command type: ${type}`);
        }
        return factory.deserialize(data);
    }

    static serialize(command: Command): { type: string; data: any } {
        const serializableCommand = command as unknown as SerializableCommand;
        if (typeof serializableCommand.serialize !== 'function') {
            throw new Error(`Command ${command.constructor.name} is not serializable`);
        }
        return {
            type: command.constructor.name,
            data: serializableCommand.serialize()
        };
    }
}

// Factory implementations
const AddShapeCommandFactory: CommandFactory = {
    deserialize(data: { shape: Shape; id: string; timestamp: number }): AddShapeCommand {
        return new AddShapeCommand(data.shape, data.id);
    }
};

const RemoveShapeCommandFactory: CommandFactory = {
    deserialize(data: { shape: Shape; id: string; timestamp: number }): RemoveShapeCommand {
        return new RemoveShapeCommand(data.shape, data.id);
    }
};

const PanCommandFactory: CommandFactory = {
    deserialize(data: { dx: number; dy: number; id: string; timestamp: number }): PanCommand {
        return new PanCommand(data.dx, data.dy, data.id);
    }
};

const UndoCommandFactory: CommandFactory = {
    deserialize(data: { commandId: string; id: string; timestamp: number }): UndoCommand {
        return new UndoCommand(data.commandId, data.id);
    }
};

const RedoCommandFactory: CommandFactory = {
    deserialize(data: { commandId: string; id: string; timestamp: number }): RedoCommand {
        return new RedoCommand(data.commandId, data.id);
    }
};

const DeleteShapeCommandFactory: CommandFactory = {
    deserialize(data: { shapeIds: string[]; deletedShapes: Shape[]; id: string; timestamp: number }): DeleteShapeCommand {
        const command = new DeleteShapeCommand(data.shapeIds, data.id);
        // Restore the deleted shapes for proper undo functionality
        (command as any).deletedShapes = data.deletedShapes;
        return command;
    }
};

// Register all command factories
CommandRegistry.register('AddShapeCommand', AddShapeCommandFactory);
CommandRegistry.register('RemoveShapeCommand', RemoveShapeCommandFactory);
CommandRegistry.register('PanCommand', PanCommandFactory);
CommandRegistry.register('UndoCommand', UndoCommandFactory);
CommandRegistry.register('RedoCommand', RedoCommandFactory);
CommandRegistry.register('DeleteShapeCommand', DeleteShapeCommandFactory);
