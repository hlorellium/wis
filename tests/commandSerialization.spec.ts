/**
 * Command Serialization Tests
 * Uses snapshots to freeze the command protocol and detect breaking changes
 */

import { describe, it, expect, beforeEach, vi } from 'vitest';
import { MoveShapesCommand, MoveVertexCommand, DeleteShapeCommand } from '../src/commands';
import { CommandRegistry } from '../src/sync/commandRegistry';
import { createMoveShapesCommand, createMoveVertexCommand, createDeleteShapeCommand } from './factories/commandFactory';
import '../src/sync/commandRegistry'; // Ensure registry is populated

// Fixed timestamp for consistent snapshots
const FIXED_TIMESTAMP = 1700000000000;

describe('Command Serialization Protocol', () => {
  beforeEach(() => {
    // Mock Date.now to return fixed timestamp for consistent snapshots
    vi.spyOn(Date, 'now').mockReturnValue(FIXED_TIMESTAMP);
  });

  describe('MoveShapesCommand', () => {
    it('should serialize consistently', () => {
      const command = createMoveShapesCommand({
        shapeIds: ['shape-1', 'shape-2'],
        deltaX: 25,
        deltaY: -10,
        id: 'test-move-shapes-123'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('move-shapes-command');
    });

    it('should serialize with negative deltas', () => {
      const command = createMoveShapesCommand({
        shapeIds: ['negative-test'],
        deltaX: -100,
        deltaY: -50,
        id: 'negative-move-123'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('move-shapes-negative');
    });

    it('should serialize with multiple shapes', () => {
      const command = createMoveShapesCommand({
        shapeIds: ['rect-1', 'rect-2', 'circle-1', 'bezier-1'],
        deltaX: 0,
        deltaY: 15,
        id: 'multi-shape-move'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('move-shapes-multiple');
    });
  });

  describe('MoveVertexCommand', () => {
    it('should serialize consistently', () => {
      const command = createMoveVertexCommand({
        shapeId: 'bezier-test',
        vertexIndex: 2,
        oldPosition: { x: 100.5, y: 200.25 },
        newPosition: { x: 125.75, y: 175.5 },
        id: 'test-move-vertex-456'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('move-vertex-command');
    });

    it('should serialize with fractional coordinates', () => {
      const command = createMoveVertexCommand({
        shapeId: 'precision-test',
        vertexIndex: 0,
        oldPosition: { x: 0.123456789, y: -0.987654321 },
        newPosition: { x: 999.999999, y: -1000.000001 },
        id: 'precision-vertex'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('move-vertex-precision');
    });

    it('should serialize with different vertex indices', () => {
      const command = createMoveVertexCommand({
        shapeId: 'vertex-index-test',
        vertexIndex: 3,
        oldPosition: { x: 50, y: 60 },
        newPosition: { x: 70, y: 80 },
        id: 'vertex-index-test'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('move-vertex-index-3');
    });
  });

  describe('DeleteShapeCommand', () => {
    it('should serialize consistently', () => {
      const command = createDeleteShapeCommand({
        shapeIds: ['delete-test-1'],
        id: 'test-delete-789'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('delete-shape-command');
    });

    it('should serialize with multiple shapes', () => {
      const command = createDeleteShapeCommand({
        shapeIds: ['shape-a', 'shape-b', 'shape-c'],
        id: 'multi-delete-test'
      });

      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized).toMatchSnapshot('delete-shapes-multiple');
    });
  });

  describe('Round-trip serialization', () => {
    it('should maintain command integrity through serialize/deserialize cycle', () => {
      const originalCommand = createMoveShapesCommand({
        shapeIds: ['roundtrip-test'],
        deltaX: 42,
        deltaY: -17,
        id: 'roundtrip-123'
      });

      // Serialize
      const serialized = CommandRegistry.serialize(originalCommand);
      
      // Deserialize
      const deserializedCommand = CommandRegistry.deserialize(serialized.type, serialized.data);
      
      // Re-serialize to compare
      const reseralized = CommandRegistry.serialize(deserializedCommand);
      
      expect(reseralized).toEqual(serialized);
    });

    it('should preserve vertex command precision', () => {
      const originalCommand = createMoveVertexCommand({
        shapeId: 'precision-roundtrip',
        vertexIndex: 1,
        oldPosition: { x: 3.14159265359, y: 2.71828182846 },
        newPosition: { x: 1.41421356237, y: 1.73205080757 },
        id: 'math-constants'
      });

      const serialized = CommandRegistry.serialize(originalCommand);
      const deserializedCommand = CommandRegistry.deserialize(serialized.type, serialized.data);
      const reseralized = CommandRegistry.serialize(deserializedCommand);
      
      expect(reseralized).toEqual(serialized);
    });
  });

  describe('Protocol versioning', () => {
    it('should include command type in serialization', () => {
      const command = createMoveShapesCommand({ id: 'type-test' });
      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized.type).toBe('MoveShapesCommand');
      expect(typeof serialized.data).toBe('object');
    });

    it('should handle command IDs consistently', () => {
      const fixedId = 'consistent-id-test-12345';
      const command = createMoveVertexCommand({ id: fixedId });
      const serialized = CommandRegistry.serialize(command);
      
      expect(serialized.data.id).toBe(fixedId);
    });

    it('should preserve timestamp fields', () => {
      const command = createDeleteShapeCommand({ id: 'timestamp-test' });
      
      // Commands should have timestamps
      expect(command.timestamp).toBeTypeOf('number');
      expect(command.timestamp).toBeGreaterThan(0);
      
      const serialized = CommandRegistry.serialize(command);
      expect(serialized.data.timestamp).toBe(command.timestamp);
    });
  });
});
