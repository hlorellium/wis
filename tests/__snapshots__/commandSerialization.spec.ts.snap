// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Command Serialization Protocol > DeleteShapeCommand > should serialize consistently > delete-shape-command 1`] = `
{
  "data": {
    "deletedShapes": [],
    "id": "test-delete-789",
    "shapeIds": [
      "delete-test-1",
    ],
    "timestamp": 1700000000000,
  },
  "type": "DeleteShapeCommand",
}
`;

exports[`Command Serialization Protocol > DeleteShapeCommand > should serialize with multiple shapes > delete-shapes-multiple 1`] = `
{
  "data": {
    "deletedShapes": [],
    "id": "multi-delete-test",
    "shapeIds": [
      "shape-a",
      "shape-b",
      "shape-c",
    ],
    "timestamp": 1700000000000,
  },
  "type": "DeleteShapeCommand",
}
`;

exports[`Command Serialization Protocol > MoveShapesCommand > should serialize consistently > move-shapes-command 1`] = `
{
  "data": {
    "dx": 25,
    "dy": -10,
    "id": "test-move-shapes-123",
    "shapeIds": [
      "shape-1",
      "shape-2",
    ],
    "timestamp": 1700000000000,
  },
  "type": "MoveShapesCommand",
}
`;

exports[`Command Serialization Protocol > MoveShapesCommand > should serialize with multiple shapes > move-shapes-multiple 1`] = `
{
  "data": {
    "dx": 0,
    "dy": 15,
    "id": "multi-shape-move",
    "shapeIds": [
      "rect-1",
      "rect-2",
      "circle-1",
      "bezier-1",
    ],
    "timestamp": 1700000000000,
  },
  "type": "MoveShapesCommand",
}
`;

exports[`Command Serialization Protocol > MoveShapesCommand > should serialize with negative deltas > move-shapes-negative 1`] = `
{
  "data": {
    "dx": -100,
    "dy": -50,
    "id": "negative-move-123",
    "shapeIds": [
      "negative-test",
    ],
    "timestamp": 1700000000000,
  },
  "type": "MoveShapesCommand",
}
`;

exports[`Command Serialization Protocol > MoveVertexCommand > should serialize consistently > move-vertex-command 1`] = `
{
  "data": {
    "id": "test-move-vertex-456",
    "newPos": {
      "x": 125.75,
      "y": 175.5,
    },
    "oldPos": {
      "x": 100.5,
      "y": 200.25,
    },
    "shapeId": "bezier-test",
    "timestamp": 1700000000000,
    "vertexIndex": 2,
  },
  "type": "MoveVertexCommand",
}
`;

exports[`Command Serialization Protocol > MoveVertexCommand > should serialize with different vertex indices > move-vertex-index-3 1`] = `
{
  "data": {
    "id": "vertex-index-test",
    "newPos": {
      "x": 70,
      "y": 80,
    },
    "oldPos": {
      "x": 50,
      "y": 60,
    },
    "shapeId": "vertex-index-test",
    "timestamp": 1700000000000,
    "vertexIndex": 3,
  },
  "type": "MoveVertexCommand",
}
`;

exports[`Command Serialization Protocol > MoveVertexCommand > should serialize with fractional coordinates > move-vertex-precision 1`] = `
{
  "data": {
    "id": "precision-vertex",
    "newPos": {
      "x": 999.999999,
      "y": -1000.000001,
    },
    "oldPos": {
      "x": 0.123456789,
      "y": -0.987654321,
    },
    "shapeId": "precision-test",
    "timestamp": 1700000000000,
    "vertexIndex": 0,
  },
  "type": "MoveVertexCommand",
}
`;
